internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -100,159 +111,217 @@ func (r *Reconciler) getNewMachineSet(ctx context.Context, d *clusterv1.MachineD
 
-	// Add foregroundDeletion finalizer to MachineSet if the MachineDeployment has it
-	finalizerSet := sets.Set[string]{}.Insert(d.Finalizers...)
-	if finalizerSet.Has(metav1.FinalizerDeleteDependents) {
-		controllerutil.AddFinalizer(&newMS, metav1.FinalizerDeleteDependents)
+	// Keep trying to get the MS. This will force the cache to update and prevent any future reconciliation of the
+	// MachineDeployment to reconcile with an outdated list of MachineSets which could lead to unwanted creation of a
+	// duplicate MachineSet.
+	if err := wait.PollImmediate(stateConfirmationInterval, stateConfirmationTimeout, func() (bool, error) {
+		ms := &clusterv1.MachineSet{}
+		if err := r.Client.Get(ctx, client.ObjectKeyFromObject(newMS), ms); err != nil {
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		return true, nil
+	}); err != nil {
+		return nil, err
 	}
 
-	allMSs := append(oldMSs, &newMS)
-	newReplicasCount, err := mdutil.NewMSNewReplicas(d, allMSs, &newMS)
+	err = r.updateMachineDeployment(ctx, d, func(innerDeployment *clusterv1.MachineDeployment) {
+		mdutil.SetDeploymentRevision(d, newRevision)
+	})
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "failed to update revision annotation on MachineDeployment")
 	}
 
-	*(newMS.Spec.Replicas) = newReplicasCount
-
-	// Set new machine set's annotation
-	mdutil.SetNewMachineSetAnnotations(d, &newMS, newRevision, false, log)
-	// Create the new MachineSet. If it already exists, then we need to check for possible
-	// hash collisions. If there is any other error, we need to report it in the status of
-	// the Deployment.
-	alreadyExists := false
-	err = r.Client.Create(ctx, &newMS)
-	createdMS := &newMS
-	switch {
-	// We may end up hitting this due to a slow cache or a fast resync of the Deployment.
-	case apierrors.IsAlreadyExists(err):
-		alreadyExists = true
+	return newMS, nil
+}
 
-		ms := &clusterv1.MachineSet{}
-		msErr := r.Client.Get(ctx, client.ObjectKey{Namespace: newMS.Namespace, Name: newMS.Name}, ms)
-		if msErr != nil {
-			return nil, msErr
+func (r *Reconciler) computeDesiredMachineSet(deployment *clusterv1.MachineDeployment, existingMS *clusterv1.MachineSet, oldMSs []*clusterv1.MachineSet, newRevision string) (*clusterv1.MachineSet, error) {
+	desiredMS := &clusterv1.MachineSet{}
+
+	var machineSetName string
+	var machineSetUID types.UID
+	var machineUniqueIdentifier string
+	var machineTemplateHash string
+	var selectorKey, selectorValue string
+	labels := map[string]string{}
+	replicas := new(int32)
+
+	if existingMS == nil {
+		// Creating a new MachineSet
+		machineUniqueIdentifier = apirand.String(10)
+		labels[clusterv1.MachineUniqueIdentifier] = machineUniqueIdentifier
+		selectorKey = clusterv1.MachineUniqueIdentifier
+		selectorValue = machineUniqueIdentifier
+
+		machineSetName = deployment.Name + "-" + apirand.SafeEncodeString(machineUniqueIdentifier)
```

Comment:
Could we hit the max length here?

Reply:
I think we could also use GenerateName here right?

We get back the new name after the SSA apply call afaik